### PR TITLE
[numpy] `torch.erfinv`: promote integer inputs to float

### PR DESCRIPTION
**Pull request by kshitij12345**
Reference: #42515

### CODE REVIEW

**Comment by kshitij12345** [comment]
**Note**

Test failures are due to the change of behaviour of `erfinv` in SciPy

`1.4.0` onwards, `scipy.special.erfinv` returns `nan` for out of domain values.
```python
>>> import scipy.special
>>> scipy.__version__
'1.4.0'
>>> scipy.special.erfinv(1.1)
nan
>>> scipy.special.erfinv(-1.1)
nan
```

Previously it used to return `inf`.
```python
>>> import scipy.special
>>> scipy.__version__
'1.3.0'
>>> scipy.special.erfinv(1.1)
inf
>>> scipy.special.erfinv(-1.1)
-inf
```

The behaviour of Pytorch matches the latest versions i.e. `nan` is returned for out of domain values.

**Comment by mruberry** [comment]
Hey @kshitij12345! Overall this looks good. I made a suggestion about refactoring the version check to make it clearer and a more usable tool.

File: torch/testing/_internal/common_methods_invocations.py [hunk 1]
```diff
@@ -27,6 +27,14 @@
 if TEST_SCIPY:
     import scipy.special
 
+
+def versiontuple(v):
```
[the remark this answers is not in the document]
Comment (kshitij12345) [reply torch/testing/_internal/common_methods_invocations.py:31]: Sure. Will move the helper.

Also, comparison of two tuple is left to right and short-circuits at the first difference.
https://stackoverflow.com/a/60303999/5602957

**Comment by kshitij12345** [comment]
@mruberry Failure looks to be a false positive. PTAL:) Thanks!

File: torch/testing/_internal/common_methods_invocations.py [hunk 2]
```diff
@@ -863,6 +863,25 @@ def reference_sigmoid(x):
...
+        UnaryUfuncInfo('erfinv',
+                       ref=scipy.special.erfinv,
+                       decorators=(precisionOverride({torch.float16: 1e-2,
+                                                      torch.bfloat16: 1e-2,
+                                                      torch.float32: 1e-4}),),
+                       dtypes=all_types_and(torch.bool),
+                       dtypesIfCPU=all_types_and(torch.bool, torch.bfloat16),
+                       dtypesIfCUDA=all_types_and(torch.bool, torch.half),
+                       promotes_integers_to_float=True,
+                       handles_extremals=False,
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:875]: Interesting. Would you file an issue for erfinv not handling extremal values properly and cite it here?

File: torch/testing/_internal/common_methods_invocations.py [hunk 3]
```diff
@@ -863,6 +863,25 @@ def reference_sigmoid(x):
...
+                       handles_large_floats=False,
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:876]: An issue for failing to handle large values (and corresponding citation here) would be great, too

**Review by mruberry — APPROVED** [review]
Looks really good, thanks @kshitij12345! Would you add the two requested comments and ping me that this is ready to merge?

File: torch/testing/_internal/common_methods_invocations.py (hunk 2 above)
Comment (kshitij12345) [reply torch/testing/_internal/common_methods_invocations.py]: Turns out that it is fine if we set both `handles_extremals` and `handles_large_float` to true. Had set them to `False` while debugging for the SciPy discrepancy.
Have updated them again!

Thanks!

**Comment by kshitij12345** [comment]
@mruberry Gentle Ping :)

File: torch/testing/_internal/common_methods_invocations.py [hunk 4]
```diff
@@ -863,6 +863,23 @@ def reference_sigmoid(x):
...
+                       dtypesIfCPU=all_types_and(torch.bool, torch.bfloat16),
+                       dtypesIfCUDA=all_types_and(torch.bool, torch.half),
+                       promotes_integers_to_float=True,
+                       domain=(-1, 1),
+                       skips=(
+                           # Reference: https://github.com/pytorch/pytorch/pull/49155#issuecomment-742664611
+                           SkipInfo('TestUnaryUfuncs', 'test_reference_numerics',
+                                    active_if=not version_atleast(scipy.__version__, required_version="1.4.0")),
+                           # RuntimeError: "pow" not implemented for 'BFloat16'
+                           SkipInfo('TestCommon', 'test_variant_consistency_jit',
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:881]: I think we can get rid of this skip now? @kshitij12345 let's accept this as now but would you publish a follow-up PR that eliminates unnecessary bfloat16 skips on test_variant_consistency? I think @zasdfgbnm made that test work properly for bfloat16.

File: torch/testing/_internal/common_utils.py [hunk 5]
```diff
@@ -1883,3 +1883,14 @@ def _assertGradAndGradgradChecks(test_case, apply_fn, inputs):
                       torch.double: 1e-5,
                       torch.half: 1e-2,
                       torch.bfloat16: 1e-1}
+
+# Returns True if version is at least required version, False otherwise
+def version_atleast(version, required_version):
```
Comment (mruberry) [inline torch/testing/_internal/common_utils.py:1888]: I looked more into this at it looks like PyTorch's pattern is to use LooseVersion for this, see:

https://github.com/pytorch/pytorch/blob/2dff0b3e918530719f7667cb31541f036a25e3f2/tools/setup_helpers/cmake.py#L119

and

https://github.com/pytorch/pytorch/blob/47c65f82231f189f40430b9c9e5b1fde4f7715ef/test/test_spectral_ops.py#L107

That should let us kill this helper and avoid the Stack Overflow citation.

File: torch/testing/_internal/common_methods_invocations.py (hunk 4 above)
Comment (kshitij12345) [reply torch/testing/_internal/common_methods_invocations.py:881]: SUre. Will do that.

Thanks!

**Comment by mruberry** [comment]
Hey @kshitij12345!

I had a chance to look at versioning some more and it looks like PyTorch's pattern is to use LooseVersion, so let's do that. Also, as mentioned in an inline comment, after this PR would you look at removing excessive bfloat16 skips for the jit variant test? I think those tests will pass now.

File: torch/testing/_internal/common_utils.py (hunk 5 above)
Comment (kshitij12345) [reply torch/testing/_internal/common_utils.py:1888]: Oh. Great! Thanks!

**Comment by kshitij12345** [comment]
@mruberry Have addressed the review :)

**Comment by mruberry** [comment]
Unfortunately the test failures real. Maybe this needs a rebase or maybe those tests do still need to be skipped?

**Comment by kshitij12345** [comment]
Looks like the BFloat16 Skip for pow is still needed.

https://circleci.com/gh/pytorch/pytorch/9780727

`RuntimeError: "pow" not implemented for 'BFloat16' `

**Comment by mruberry** [comment]
> Looks like the BFloat16 Skip for pow is still needed.
> 
> https://circleci.com/gh/pytorch/pytorch/9780727
> 
> `RuntimeError: "pow" not implemented for 'BFloat16' `

Thanks for trying it. Sorry I was mistaken about this.

**Comment by kshitij12345** [comment]
@mruberry PTAL :)

Failure looks irrelevant

```
 test_ReplicationPad1d_alert_nondeterministic_cuda (__main__.TestNN) ... ok (0.018s)
  test_ReplicationPad1d_cuda (__main__.TestNN) ... Traceback (most recent call last):
  File "run_test.py", line 910, in <module>
    main()
  File "run_test.py", line 889, in main
    raise RuntimeError(err_message)
RuntimeError: test_nn failed!

(base) circleci@PACKER-5FD865C5 C:\Users\circleci\project\test>if ERRORLEVEL 1 exit /b 1 
+ cleanup
+ retcode=1
+ set +x
```

https://app.circleci.com/pipelines/github/pytorch/pytorch/254329/workflows/95ec199b-9080-4c03-848d-8a4fd9e52cd0/jobs/9806399

**Comment by mruberry** [comment]
Hey @kshitij12345, would you rebase this, please? Sorry about all the merge conflicts. As more OpInfos are added there should be fewer of them.

**Pull request closed without merging**